core/src/main/java/bisq/core/btc/nodes/LocalBitcoinNode.java

```diff
@@ -25,39 +54,321 @@
+     * a ready-made implementation.
+     */
+    private Optional<VersionMessage> attemptHandshakeForVersionMessage() {
+        Peer peer;
+        try {
+            peer = createLocalPeer(port);
+        } catch (UnknownHostException ex) {
+            log.error("Local bitcoin node handshake attempt unexpectedly threw: {}", ex);
+            return Optional.empty();
+        }
+
+        /* We temporarily silence BitcoinJ NioClient's and NioClientManager's loggers,
+         * because when a local Bitcoin node is not found they pollute console output
+         * with "connection refused" error messages.
+         */
+        var originalNioClientLoggerLevel = silence(NioClient.class);
+        var originalNioClientManagerLoggerLevel = silence(NioClientManager.class);
+
+        NioClient client;
+
+        try {
+            log.info("Initiating attempt to connect to and handshake with a local "
+                    + "Bitcoin node (which may or may not be running) on port {}.",
+                    port);
+            client = createClient(peer, port, CONNECTION_TIMEOUT);
         } catch (IOException ex) {
-            log.info("No local Bitcoin node detected on port {}.", port);
+            log.error("Local bitcoin node handshake attempt unexpectedly threw: {}", ex);
+            return Optional.empty();
         }
-        callback.run();
+
+        ListenableFuture<VersionMessage> peerVersionMessageFuture = getVersionMessage(peer);
+
+        Optional<VersionMessage> optionalPeerVersionMessage;
+
+        // block for VersionMessage or cancellation (in case of connection failure)
+        try {
+            var peerVersionMessage = peerVersionMessageFuture.get();
+            optionalPeerVersionMessage = Optional.of(peerVersionMessage);
+        } catch (ExecutionException | InterruptedException | CancellationException ex) {
+            optionalPeerVersionMessage = Optional.empty();
+        }
+
+        peer.close();
+        client.closeConnection();
+
+        restoreLoggerLevel(NioClient.class, originalNioClientLoggerLevel);
+        restoreLoggerLevel(NioClientManager.class, originalNioClientManagerLoggerLevel);
+
+        return optionalPeerVersionMessage;
+    }
+
+    private static Level silence(Class klass) {
+        var logger = getLogger(klass);
+        var originalLevel = logger.getLevel();
+        logger.setLevel(Level.OFF);
+        return originalLevel;
+    }
+
+    private static void restoreLoggerLevel(Class klass, Level originalLevel) {
+        getLogger(klass).setLevel(originalLevel);
+    }
+
+    private static ch.qos.logback.classic.Logger getLogger(Class klass) {
+        return (ch.qos.logback.classic.Logger) LoggerFactory.getLogger(klass);
+    }
+
+    private ListenableFuture<VersionMessage> getVersionMessage(Peer peer) {
+        SettableFuture<VersionMessage> peerVersionMessageFuture = SettableFuture.create();
+
```

Comment:
Blank line

Reply:
Here rationale is to separate the creation of the future from the parts that use it. The next three parts of the method operate on this future.